### DISCUSSION TITLE
no-undef with unnecessary alerts?

### DISCUSSION THREAD

**Discussion by HiranChaudhuri** [eslint/eslint#17858] [Q&A]
I just started applying eslint to an open source project. That project contains JavaScript files that seemingly have worked since quite some years. Now eslint is coming up with a lot of findings - so far so good. But it is the no-undef rule that makes us think.

Here is an excerpt of the script. It configures an event handler for the embedded JavaScript engine:

```
this.aiStarted = function() {
	this.ai = new worldScripts["oolite-libPriorityAI"].PriorityAIController(this.ship);
	ai.setParameter("oolite_flag_watchForCargo",true);
	ai.setParameter("oolite_personalityMatchesLeader",0.9);
        ...
}
```

We get to see many alarms like

```
   34:2   error  'ai' is not defined  no-undef
   36:2   error  'ai' is not defined  no-undef

```

Now we'd like to understand whether `this.ai` and `ai` are so different and this has to be corrected, or if eslint by accident (misconfiguration or bug) does not spot the resemblance. If it can be configured what would we have to do? 

The [documentation](https://eslint.org/docs/latest/rules/no-undef) does not help in this respect.

**Comment by nzakas** [comment]
`ai` is a variable and `this.ai` is a property, and ESLint can't determine if they are the same.

The only way this code works is if `this` is actually `globalThis`, in which case assigning `this.ai = "foo"` creates a global variable `ai` that you are reading later.

So the rule is correct. You should either use `this.ai` everywhere or use `ai` everywhere to avoid ambiguity.
